src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java

```diff
@@ -190,10 +208,142 @@ public ResultSet getClientInfoProperties() throws SQLException {
         throw new SQLFeatureNotSupportedException("TODO");
     }
 
+    private MongoJsonSchema getFunctionJsonSchema() {
+        MongoJsonSchema schema = new MongoJsonSchema();
+        schema.bsonType = "object";
+        schema.additionalProperties = false;
+        schema.required = new HashSet<String>();
+        schema.properties = new HashMap<String, MongoJsonSchema>();
+
+        schema.required.add(FUNCTION_CAT);
+        schema.required.add(FUNCTION_SCHEM);
+        schema.required.add(FUNCTION_NAME);
+        schema.required.add(REMARKS);
+        schema.required.add(FUNCTION_TYPE);
+        schema.required.add(SPECIFIC_NAME);
+
+        schema.properties.put(FUNCTION_CAT, stringSchema);
+        schema.properties.put(FUNCTION_SCHEM, stringSchema);
+        schema.properties.put(FUNCTION_NAME, stringSchema);
+        schema.properties.put(REMARKS, stringSchema);
+        schema.properties.put(FUNCTION_TYPE, intSchema);
+        schema.properties.put(SPECIFIC_NAME, stringSchema);
+        return schema;
+    }
+
+    private BsonDocument getFunctionValuesDoc(String functionName, String remarks) {
+        BsonDocument root = new BsonDocument();
+        BsonDocument bot = new BsonDocument();
+        root.put("", bot);
+        bot.put(FUNCTION_CAT, new BsonString("def"));
+        bot.put(FUNCTION_SCHEM, new BsonNull());
+        bot.put(FUNCTION_NAME, new BsonString(functionName));
+        bot.put(REMARKS, new BsonString(remarks));
+        bot.put(FUNCTION_TYPE, new BsonInt32(functionNoTable));
+        bot.put(SPECIFIC_NAME, new BsonString(functionName));
+        return root;
+    }
+
     @Override
     public ResultSet getFunctions(String catalog, String schemaPattern, String functionNamePattern)
             throws SQLException {
-        throw new SQLFeatureNotSupportedException("TODO");
+        ArrayList<BsonDocument> docs = new ArrayList<>(MongoFunction.mongoSQLFunctionNames.length);
+        MongoJsonSchema schema = getFunctionJsonSchema();
+
+        Pattern functionPatternRE = null;
+        if (functionNamePattern != null) {
+            functionPatternRE = Pattern.compile(functionNamePattern.replaceAll("%", ".*"));
```

Comment:
I'm working with patterns in my PR too and wonder if this is sufficient. Don't we also need to replace `_` with `.`, and should we escape other java-style regex characters that otherwise likely aren't escaped in sql-style patterns?

Reply:
Probably, just copied this from what we did in `MySQL`. You think we want to escape any escape sequence?

Reply:
I think this gets tricky. Let's support `_` and make another ticket for possible regex characters.